ls1intum/Artemis
# [Enhancement] AssessmentDueDate for programming exercises

### Checklist

### Motivation and Context


The assessment due date was not fully implemented for programming exercises. Furthermore, this solves also an issue which we encountered during testing of #2131. Student could see in the response of the REST call the result of his submission even though he should not be able to as the assessment due date was not passed.

### Description

- Added check for assessmentDueDate in `findResultsFilteredForStudents` for the exercise `/details` REST call
- Implemented override for `findResultsFilteredForStudents` for programming exercises `/for-dashboard` REST call
- Implemented override for `findLatestSubmissionWithRatedResultWithCompletionDate` 
- Adapted database query to check for assessmentDueDate for the Student CodeEditor
- Added label to indicate that the assessment due date has passed and that the results will be published immediately (when assessing a exercise for the first time after the assessment due date)
- Remove result button on scores page when assessment due date is in the past (programming exercises)

### Steps for Testing


**Alert and functionality in assessment view when assessment due date is in the past**
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a exercise (Programming, File Upload, Modeling and Text)
4. Participate on exercise
5. Set Assessment Due Date
6. Check if alert is there which indicates that first assessment of submission will be published immediately when assessment due date is in the past
7. Check (as tutor) that after you clicked on submit the override button is not displayed anymore (as instructor you can override as usual)
8. Go to the scores page of the programming exercise, there the result button should not be displayed anymore when the assessment due date is in the past

**Publishing of results before assessment due date and after assessment due date (student)**
1. Log in to Artemis as tutor/instructor
2. Navigate to Course Administration
3. Remove assessment due date from created exercises and assess the submissions
4. Login as student, but keep logged in as well as tutor/instructor
5. Go to course overview -> select your course -> check if you can see the assessed result (course-exercise view + course exercise detail view + student submission view (e.g. Student Code Editor)
6. Check that the assessed result is displayed
7. Now set a assessment due date in the future
8. Check that for Modeling, FileUpload and Text no result is displayed at all. For Programming you should see the automatic result
9. Set the assessment due date now in the past
10. Check that for all exercises the latest manual result is displayed.
11. Check also network tab for the three views, that no result is send out before the assessment due date :) 


### Test Coverage

* ExerciseService.java: 85%
* programming-exercise.component.ts 95%

### Screenshots


**Alert and functionality in assessment view when assessment due date is in the past (tutor/instructor)**

Alert that indicates that the first assessment of a submission will be published immediately:
<img width="1625" alt="image" src="https://user-images.githubusercontent.com/41108487/95241840-a5d9a280-080e-11eb-816f-9244a38d4270.png">

No override button when assessment due date is in the past:
<img width="1669" alt="image" src="https://user-images.githubusercontent.com/41108487/95242538-9ad34200-080f-11eb-90f9-ad12dc9fcbf9.png">

Result button on scores pages before assessment due date:
<img width="1623" alt="image" src="https://user-images.githubusercontent.com/41108487/95242895-1d5c0180-0810-11eb-979a-7b30a196dd4a.png">

..and after assessment due date:
<img width="1626" alt="image" src="https://user-images.githubusercontent.com/41108487/95243029-51cfbd80-0810-11eb-9700-f8376f0a93a3.png">


**Publishing of results before assessment due date and after assessment due date (student)**

Before assessment due date - Course exercise view:
<img width="1138" alt="image" src="https://user-images.githubusercontent.com/41108487/95248380-82ffbc00-0817-11eb-839f-8e2d04de1a42.png">

Before assessment due date - Course exercise detail view:
<img width="1669" alt="image" src="https://user-images.githubusercontent.com/41108487/95248473-a0cd2100-0817-11eb-9796-62e54718f03e.png">

Before assessment due date - Student submission view:
<img width="1640" alt="image" src="https://user-images.githubusercontent.com/41108487/95248520-b0e50080-0817-11eb-90ec-fc611109197d.png">


After assessment due date - Course exercise view:
<img width="1253" alt="image" src="https://user-images.githubusercontent.com/41108487/95248863-2f41a280-0818-11eb-8313-11889ae36145.png">
After assessment due date - Course exercise detail view:
<img width="1660" alt="image" src="https://user-images.githubusercontent.com/41108487/95248829-2224b380-0818-11eb-8e38-1d1d93a33aab.png">

After assessment due date - Student submission view:
<img width="1644" alt="image" src="https://user-images.githubusercontent.com/41108487/95248777-16d18800-0818-11eb-8de2-2c8f4c212a4d.png">
